rotkehlchen/assets/asset.py

```diff
@@ -1129,3 +1132,89 @@ def from_identifier(
             )
         except DeserializationError:
             return None
+
+
+# Create a generic variable that can be 'CustomAsset', or any subclass such as stock, painting, etc.  # noqa: E501
+R = TypeVar('R', bound='CustomAsset')
+
+
+@dataclass(init=True, repr=True, eq=False, order=False, unsafe_hash=False, frozen=True)
+class CustomAsset(Asset):
+    """Class to represent assets added by users that are not crypto coins"""
+    custom_asset_type: str = field(init=False)
+    notes: Optional[str] = field(init=False)
+
+    @classmethod
+    def initialize(  # type: ignore  # figure out a way to make mypy happy
+            cls: Type[R],
+            identifier: str,
+            name: str,
+            custom_asset_type: Optional[str] = None,
+            notes: Optional[str] = None,
+            swapped_for: Optional[Asset] = None,
+            started: Optional[Timestamp] = None,
+    ) -> R:
+        """Initialize a custom asset from fields"""
+        custom_asset = cls('whatever', direct_field_initialization=True)
+        object.__setattr__(custom_asset, 'identifier', identifier)
+        object.__setattr__(custom_asset, 'name', name)
+        object.__setattr__(custom_asset, 'symbol', None)
+        object.__setattr__(custom_asset, 'asset_type', AssetType.CUSTOM_ASSET)
+        object.__setattr__(custom_asset, 'custom_asset_type', custom_asset_type)
+        object.__setattr__(custom_asset, 'notes', notes)
+        object.__setattr__(custom_asset, 'started', started)
+        object.__setattr__(custom_asset, 'forked', None)
+        object.__setattr__(custom_asset, 'swapped_for', swapped_for)
+        object.__setattr__(custom_asset, 'cryptocompare', None)
+        object.__setattr__(custom_asset, 'coingecko', None)
+        return custom_asset
+
+    @classmethod
+    def from_identifier(
+            cls: Type[R],
+            identifier: str,
+            form_with_incomplete_data: bool = True,
+    ) -> Optional[R]:
+        """Attempts to turn an asset into a `CustomAsset` object. If it fails returns None"""
+        try:
+            return cls(
+                identifier,
+                form_with_incomplete_data=form_with_incomplete_data,
+            )
+        except DeserializationError:
+            return None
+
+    @classmethod
+    def deserialize_from_db(
+            cls: Type[R],
+            entry: Tuple[str, str, Optional[str], str, Optional[int], Optional[str]],
+    ) -> R:
+        """Takes a `custom_asset` entry from DB and turns it into a `CustomAsset` instance."""
+        return cls.initialize(
+            identifier=entry[0],
+            name=entry[1],
+            notes=entry[2],
+            custom_asset_type=entry[3],
+            started=Timestamp(int(entry[4])) if entry[4] is not None else None,
+            swapped_for=cls.from_identifier(entry[5]) if entry[5] is not None else None,
```

Comment:
One thing that I just noticed is that right now when deserializing `swapped_for` we are assuming that it is a custom asset (since `CustomAsset`'s `from_identifier` is used). I think it is incorrect since `swapped_for` can be any asset. What do you think about it?